From a2b1984b6d4afb824e077df69dc3f5c58bdefb31 Mon Sep 17 00:00:00 2001
From: "T.J. Mercier" <tjmercier@google.com>
Date: Tue, 14 Apr 2026 13:18:56 -0700
Subject: [PATCH 131/223] ANDROID: read task_dma_buf_info under task lock and
 pin it

The per-process dmabuf accounting procfs files read a task's dmabuf_info
struct to compute and show RSS/PSS, but no reference to the dmabuf_info
struct is taken, so it can be freed by an exec() during
proc_dmabuf_rss_show, proc_dmabuf_rss_hwm_show,
proc_dmabuf_rss_hwm_write, and proc_dmabuf_pss_show resulting in a UAF.

Fix this by taking a temporary reference on the task's dmabuf_info in
those functions. This is performed under the task_lock() to protect
against changes to a task's dmabuf_info pointer resulting from other
operations like unshare(), exec(), or close_range().

Bug: 496996374
Fixes: 09725f5bd79d ("ANDROID: Don't use current pointer for per-process dmabuf accounting")
Change-Id: I1d93c2e0b93e393a77f09da47401e91f835dd2d1
Signed-off-by: T.J. Mercier <tjmercier@google.com>
(cherry picked from commit 9d923a8ee28fa36f28abf98ae2fdea88b1c13d2b)
Signed-off-by: Lee Jones <joneslee@google.com>
---
 drivers/dma-buf/dma-buf.c | 30 +++++++++++++++++------
 fs/file.c                 |  6 ++++-
 fs/proc/base.c            | 51 +++++++++++++++++++++++++++++----------
 3 files changed, 65 insertions(+), 22 deletions(-)

diff --git a/drivers/dma-buf/dma-buf.c b/drivers/dma-buf/dma-buf.c
index d08b98a3ffc0..43bba96e1741 100644
--- a/drivers/dma-buf/dma-buf.c
+++ b/drivers/dma-buf/dma-buf.c
@@ -599,7 +599,7 @@ static struct task_dma_buf_info *compute_dmabuf_info(struct task_struct *task)
 
 int copy_dmabuf_info(u64 clone_flags, struct task_struct *task)
 {
-	struct task_dma_buf_info *parent_dmabuf_info = current->dmabuf_info;
+	struct task_dma_buf_info *parent_dmabuf_info;
 	struct task_dma_buf_info *child_dmabuf_info;
 	bool share_vm = clone_flags & CLONE_VM;
 	bool share_fs = clone_flags & CLONE_FILES;
@@ -626,21 +626,31 @@ int copy_dmabuf_info(u64 clone_flags, struct task_struct *task)
 		return 0;
 	}
 
-	/*
-	 * Partial sharing is not supported.
-	 * Children of such tasks are also not supported.
-	 */
-	if (share_vm != share_fs || !parent_dmabuf_info) {
+	task_lock(current);
+	parent_dmabuf_info = current->dmabuf_info;
+	if (parent_dmabuf_info)
+		get_dmabuf_info(parent_dmabuf_info);
+	task_unlock(current);
+
+	/* Children of tasks with no accounting info are not supported. */
+	if (!parent_dmabuf_info) {
 		task->dmabuf_info = NULL;
 		return 0;
 	}
 
+	/* Partial sharing is not supported. */
+	if (share_vm != share_fs) {
+		task->dmabuf_info = NULL;
+		put_dmabuf_info(parent_dmabuf_info);
+		return 0;
+	}
+
 	/*
 	 * Full sharing: Both MM and FD references to dmabufs are shared with
 	 * the parent, so they can both share the same dmabuf_info.
 	 */
 	if (share_vm && share_fs) {
-		get_dmabuf_info(parent_dmabuf_info);
+		/* task takes the parent_dmabuf_info with the elevated refcount. */
 		task->dmabuf_info = parent_dmabuf_info;
 
 		if (task->mm) {
@@ -662,6 +672,7 @@ int copy_dmabuf_info(u64 clone_flags, struct task_struct *task)
 	 * the child. Construct the dmabuf accounting info for the child here based on what actually
 	 * made it into the child's files_struct and mm_struct.
 	 */
+	put_dmabuf_info(parent_dmabuf_info);
 	child_dmabuf_info = compute_dmabuf_info(task);
 	if (!child_dmabuf_info)
 		return -ENOMEM;
@@ -716,7 +727,7 @@ void put_dmabuf_info(struct task_dma_buf_info *dmabuf_info)
 int dma_buf_begin_new_exec(struct files_struct *old_files)
 {
 	struct task_dma_buf_info *new_dmabuf_info;
-	struct task_dma_buf_info *old_dmabuf_info = current->dmabuf_info;
+	struct task_dma_buf_info *old_dmabuf_info;
 	struct files_struct *my_files = current->files;
 
 	if (!static_key_enabled(&dmabuf_accounting_key))
@@ -795,7 +806,10 @@ int dma_buf_begin_new_exec(struct files_struct *old_files)
 		task_dmabuf_records_preload_end();
 	}
 
+	task_lock(current);
+	old_dmabuf_info = current->dmabuf_info;
 	current->dmabuf_info = new_dmabuf_info; // refcount from alloc_task_dma_buf_info
+	task_unlock(current);
 	put_dmabuf_info(old_dmabuf_info);
 
 	return 0;
diff --git a/fs/file.c b/fs/file.c
index 1967ef670605..159467938b38 100644
--- a/fs/file.c
+++ b/fs/file.c
@@ -758,6 +758,7 @@ int __close_range(unsigned fd, unsigned max_fd, unsigned int flags)
 
 	if ((flags & CLOSE_RANGE_UNSHARE) && atomic_read(&cur_fds->count) > 1) {
 		struct fd_range range = {fd, max_fd}, *punch_hole = &range;
+		struct task_dma_buf_info *dmabuf_info;
 
 		/*
 		 * If the caller requested all fds to be made cloexec we always
@@ -777,8 +778,11 @@ int __close_range(unsigned fd, unsigned max_fd, unsigned int flags)
 		 * the accounting info from the task. Leave the cur_fds->dmabuf_info so any existing
 		 * accounting can be unaccounted properly.
 		 */
-		put_dmabuf_info(current->dmabuf_info);
+		task_lock(current);
+		dmabuf_info = current->dmabuf_info;
 		current->dmabuf_info = NULL;
+		task_unlock(current);
+		put_dmabuf_info(dmabuf_info);
 		/*
 		 * We used to share our file descriptor table, and have now
 		 * created a private one, make sure we're using it below.
diff --git a/fs/proc/base.c b/fs/proc/base.c
index 6bc54ec7c1d8..9d9a98f0392d 100644
--- a/fs/proc/base.c
+++ b/fs/proc/base.c
@@ -3319,10 +3319,24 @@ static int proc_stack_depth(struct seq_file *m, struct pid_namespace *ns,
 #endif /* CONFIG_STACKLEAK_METRICS */
 
 #ifdef CONFIG_DMA_SHARED_BUFFER
+
+static struct task_dma_buf_info *get_task_dmabuf_info(struct task_struct *task)
+{
+	struct task_dma_buf_info *dmabuf_info;
+
+	task_lock(task);
+	dmabuf_info = task->dmabuf_info;
+	if (dmabuf_info)
+		get_dmabuf_info(dmabuf_info);
+	task_unlock(task);
+
+	return dmabuf_info;
+}
+
 static int proc_dmabuf_rss_show(struct seq_file *m, struct pid_namespace *ns,
 		     struct pid *pid, struct task_struct *task)
 {
-	struct task_dma_buf_info *dmabuf_info = task->dmabuf_info;
+	struct task_dma_buf_info *dmabuf_info = get_task_dmabuf_info(task);
 
 	if (dmabuf_info) {
 		unsigned long rss;
@@ -3330,6 +3344,7 @@ static int proc_dmabuf_rss_show(struct seq_file *m, struct pid_namespace *ns,
 		spin_lock(&dmabuf_info->lock);
 		rss = dmabuf_info->rss;
 		spin_unlock(&dmabuf_info->lock);
+		put_dmabuf_info(dmabuf_info);
 		seq_printf(m, "%lu\n", rss);
 	}
 
@@ -3340,18 +3355,22 @@ static int proc_dmabuf_rss_hwm_show(struct seq_file *m, void *v)
 {
 	struct inode *inode = m->private;
 	struct task_struct *task;
+	struct task_dma_buf_info *dmabuf_info;
 	int ret = 0;
 
 	task = get_proc_task(inode);
 	if (!task)
 		return -ESRCH;
 
-	if (task->dmabuf_info) {
+	dmabuf_info = get_task_dmabuf_info(task);
+
+	if (dmabuf_info) {
 		unsigned long rss_hwm;
 
-		spin_lock(&task->dmabuf_info->lock);
-		rss_hwm = task->dmabuf_info->rss_hwm;
-		spin_unlock(&task->dmabuf_info->lock);
+		spin_lock(&dmabuf_info->lock);
+		rss_hwm = dmabuf_info->rss_hwm;
+		spin_unlock(&dmabuf_info->lock);
+		put_dmabuf_info(dmabuf_info);
 		seq_printf(m, "%lu\n", rss_hwm);
 	}
 
@@ -3371,6 +3390,7 @@ proc_dmabuf_rss_hwm_write(struct file *file, const char __user *buf,
 {
 	struct inode *inode = file_inode(file);
 	struct task_struct *task;
+	struct task_dma_buf_info *dmabuf_info;
 	unsigned long long val;
 	int ret;
 
@@ -3385,12 +3405,15 @@ proc_dmabuf_rss_hwm_write(struct file *file, const char __user *buf,
 	if (!task)
 		return -ESRCH;
 
-	if (!task->dmabuf_info) {
+	dmabuf_info = get_task_dmabuf_info(task);
+
+	if (!dmabuf_info) {
 		ret = -ENOENT;
 	} else {
-		spin_lock(&task->dmabuf_info->lock);
-		task->dmabuf_info->rss_hwm = task->dmabuf_info->rss;
-		spin_unlock(&task->dmabuf_info->lock);
+		spin_lock(&dmabuf_info->lock);
+		dmabuf_info->rss_hwm = dmabuf_info->rss;
+		spin_unlock(&dmabuf_info->lock);
+		put_dmabuf_info(dmabuf_info);
 	}
 
 	put_task_struct(task);
@@ -3409,13 +3432,14 @@ static const struct file_operations proc_dmabuf_rss_hwm_operations = {
 static int proc_dmabuf_pss_show(struct seq_file *m, struct pid_namespace *ns,
 		     struct pid *pid, struct task_struct *task)
 {
+	struct task_dma_buf_info *dmabuf_info = get_task_dmabuf_info(task);
 	struct task_dma_buf_record *rec;
 
-	if (task->dmabuf_info) {
+	if (dmabuf_info) {
 		unsigned long pss = 0;
 
-		spin_lock(&task->dmabuf_info->lock);
-		list_for_each_entry(rec, &task->dmabuf_info->dmabufs, node) {
+		spin_lock(&dmabuf_info->lock);
+		list_for_each_entry(rec, &dmabuf_info->dmabufs, node) {
 			s64 refs = atomic64_read(&rec->dmabuf->nr_task_refs);
 
 			if (refs <= 0) {
@@ -3425,7 +3449,8 @@ static int proc_dmabuf_pss_show(struct seq_file *m, struct pid_namespace *ns,
 
 			pss += rec->dmabuf->size / (size_t)refs;
 		}
-		spin_unlock(&task->dmabuf_info->lock);
+		spin_unlock(&dmabuf_info->lock);
+		put_dmabuf_info(dmabuf_info);
 		seq_printf(m, "%lu\n", pss);
 	}
 
-- 
2.50.1

